### ISSUE TITLE
Exodatasets & Vespa Indexes

### ISSUE THREAD

**Issue by docuracy** [WorldHistoricalGazetteer/place#29]
An improved indexing system using [Vespa](https://github.com/vespa-engine/vespa?tab=readme-ov-file), together with the overhaul of the system for asserting links between Places outlined [here](https://github.com/WorldHistoricalGazetteer/place/issues/25), might remove the current need for manual reconciliation and accessioning. This approach can eliminate the need for multiple passes through contributed datasets, offering contributors a unified interface to complete all stages of place data processing in a single pass, rather than waiting for dataset-wide match-seeking operations to finish.

This system could present a Cluster of potential matches from exodatasets like Wikidata, GeoNames, and others (see below), as well as from other WHG datasets, and it might suggest relevant citations from LLMs ([proposed here](https://github.com/WorldHistoricalGazetteer/place/issues/32)). It would allow a dataset contributor to focus on a single place in their dataset and complete all stages of its processing before moving on to another. It would also reduce the friction that arises when dealing with historical places which do not exist in the modern exodatasets, but which may well be present in other contributed datasets.

## Vespa

Vespa offers hybrid search capabilities that can seamlessly combine traditional keyword-based search with vector-based search methods. Running in a Docker container, it would allow ranking based on textual, spatial, temporal, linguistic, phonetic, and semantic facets across both global and regional exodatasets _and_ contributed WHG data. Vespa supports real-time updates and provides a broad range of APIs for querying and data management.

## Normalised Place Records (NPRs)

Each exodataset would require its own transformer function to produce indexable NPRs, with namespaced `@id`s recording the authority name. The NPRs would include many of the [LPF](https://github.com/LinkedPasts/linked-places-format/blob/main/README.md) properties, with these additional fields (among others):
- Embeddings (for each toponym) for phonetic and semantic representations. These would be generated using pre-trained multilingual [G2P](https://github.com/lingjzhu/CharsiuG2P) and [BERT](https://github.com/google-research/bert/blob/master/multilingual.md) models (which would also provide embeddings for search terms).
- Simplified representations of geometries as points and bounding boxes, to be used in containment and proximity queries.
- Reduction of timespans to an array of start and end years for each record.

## Benefits

Features from WHG dataset contributions would also be transformed to NPRs (with an additional raw LPF field for swift retrieval of entire place records and datasets) and stored in Vespa, and once fully accessioned these would be available in API searches. Vespa would also deliver dataset downloads based on stored raw LPF, and dataset FeatureCollections customised for map tileset generation and efficient browser-based map visualisations.

This system would allow:

- Automatic acceptance for any "very obvious" matches in the accessioning workflow.
- Preparation of candidate-matches for entire datasets in the background while a contributor steps through individual records.
- Great improvement to the quality and extent of our existing range of APIs.
- Improved support for search incorporating toponymic [BCP 47](https://www.rfc-editor.org/rfc/bcp/bcp47.txt) language-tags.
- Replacement of our inefficient mapdata filesystem cache.
- Removal of Place data from our Postgres database, for both submitted datasets and exodatasets.
- Considerable reduction in backup data volume and overhead.

It should be implemented in such a way that the indexing of each exodataset might be periodically and independently refreshed.

## Reinforcement Learning

Integration of [Stable-Baselines3](https://stable-baselines3.readthedocs.io/en/master/) for reinforcement learning in the Vespa indexing system could further enhance the place-matching and ranking process. User interactions—such as marking good or bad matches when contributing a dataset—can be used as training data, allowing the model to adjust future rankings. The search engine would continue to refine and optimise its performance, delivering more accurate place matches both on the web site and in the WHG APIs, and reducing the need for manual intervention in the accessioning workflow.

## Proposed Exodatasets

**Further suggestions of exodatasets not already aggregated by GeoNames would be very welcome - _please add in a comment below_.**

### Global

- GeoNames (25m place names aggregated from [numerous sources](https://www.geonames.org/datasources/)).
- Wikidata: open a torrent of all entities from https://www.wikidata.org/wiki/Wikidata:Database_download with ijson, and filter places from the stream directly into Vespa.
- [OSM](https://planet.openstreetmap.org/) (6m+ nodes tagged as places).
- [Pleiades](https://pleiades.stoa.org/downloads) (63,282 toponyms; 71,082 attestations; 37,743 places).
- [TGN](https://www.getty.edu/research/tools/vocabularies/obtain/download.html) (3m+ place records; 5m+ names) "_While most records in TGN include coordinates, these coordinates are approximate and are intended for reference ("finding purposes") only (as is true of coordinates in most atlases and other resources_".
- Library of Congress ([LOC](https://authorities.loc.gov/webvoy.htm)) (_not geolocated, but linked to other geolocated sources_: [example](https://lccn.loc.gov/n93103188/marcxml)).

#### DE

- [DNB: Deutsche National Bibliothek](https://data.dnb.de/opendata/) (_Incorporates some coordinates directly from GeoNames_)

#### GB

- [EPNS: Survey of English Place-Names](https://epns.nottingham.ac.uk/about)
- [GB1900](https://www.pastplace.org/data/#tabgb1900) UK-specific dataset with over 2.5 million historical place names.
- [RCAHMW: List of Historic Place Names of Wales](https://historicplacenames.rcahmw.gov.uk/) (700k+ names).
- [Semantic Name Authority Repository Cymru](https://snarc-llgc.wikibase.cloud/wiki/Main_Page) (60k+ place names)

#### PL

- [PRNG: Państwowy Rejestr Nazw Geograficznych](https://www.geoportal.gov.pl/pl/dane/panstwowy-rejestr-nazw-geograficznych-prng/)
